### PR TITLE
Avoid planning unnecessary LIMIT/TopN/Sort/DistinctLimit

### PR DESCRIPTION
**Pull request by fgwang7w**
Cherry-pick of https://github.com/prestosql/presto/pull/441 and https://github.com/prestosql/presto/pull/818

```
== RELEASE NOTES ==

General Changes
* Avoid planning unnecessary LIMIT/TopN/Sort/DistinctLimit when relation is know to single row or less rows than requested
* The analyzer will emit a warning if a redundant ORDER BY is present
```
resolves: #14897

### CODE REVIEW

**Comment by fgwang7w** [comment]
@mbasmanova @tdcmeehan this commit is ready for review. thank you!

**Comment by fgwang7w** [comment]
Hi @rschlussel could you please help review this pr? this item is perf-related, thank you!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    @Override
+    public Pattern<SortNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(SortNode node, Captures captures, Context context)
+    {
+        if (isAtMost(node.getSource(), context.getLookup(), 0)) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java:41]: when would this e true.  this doesn't seem related to sort nodes.  Maybe it should be its own rule to replace subplans that would return zero rows with an empty values.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputVariables(), ImmutableList.of()));
+        }
+        if (isScalar(node.getSource(), context.getLookup())) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java:44]: if we change the isAtMost(0) rule to be separate, then this should be isAtMostScalar()

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public Pattern<TopNNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(TopNNode node, Captures captures, Context context)
+    {
+        if (node.getCount() == 0) {
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java:48]: same comment as above regarding zero and scalar

**Comment by rschlussel** [comment]
also, can you separate out the changes in the analyzer/planner (don't plan unnecessary sort) into its own commit separate from the new optimizers, so it's easier to see what's going on there.  I haven't looked at that part yet because it's hard to find all the related pieces.

**Comment by kaikalur** [comment]
I'm thinking this PR should be broken up into 2 or even 3. In particular, it will be good to do redundant orderby in a separate PR

**Comment by fgwang7w** [comment]
@kaikalur @rschlussel thank you all for reviewing the code changes. I am currently reworking this PR into multiple commits and put the unnecessary sort and redundant orderby in different PRs.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java (hunk 1 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java:41]: So the new rule is EvaluateZeroCount which evaluates zeroSample, zero-topN, zero-limit, and zero-distinctLimit count.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java (hunk 2 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantSort.java:44]: Since the new rule does not implement a particular planNode, my take is to remain this code as it is to lookup from the tree and can be verified by testForZeroCardinality

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java (hunk 3 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java]: this check is moved to the new rule now

**Review by rschlussel — COMMENTED** [review]
The commit "Avoid planning unnecessary sort" also has lots of formatting and similar changes that belong in the first commit.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantLimit.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public Pattern<LimitNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(LimitNode limit, Captures captures, Context context)
+    {
+        if (limit.getCount() == 0) {
```
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantLimit.java:45]: yes, this check not needed, removed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EvaluateZeroCount.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import com.facebook.presto.spi.WarningCollector;
+import com.facebook.presto.spi.plan.*;
+import com.facebook.presto.sql.planner.PlanVariableAllocator;
+import com.facebook.presto.sql.planner.TypeProvider;
+import com.facebook.presto.sql.planner.plan.SampleNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.google.common.collect.ImmutableList;
+
+public class EvaluateZeroCount
+    implements PlanOptimizer
```
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/EvaluateZeroCount.java]: ok will create new iterative rules for each case. So is it recommended not to create new PlanOplanizer for new rule? or is there any limitation on using planOptimizer?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static com.facebook.presto.sql.planner.plan.Patterns.topN;
+
+/**
+ * Replace TopN node
+ * 1. With a Sort node when the subplan is guaranteed to produce fewer rows than N
+ * 2. With its source when the subplan produces only one row
+ * 3. With a empty ValuesNode when N is 0
```
[the remark this answers is not in the document]
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java:30]: fixed

**Comment by fgwang7w** [comment]
Hi @rschlussel , I have fixed all revised changes. could you pleaes give another round of review for merging this commit? Many thanks!

**Comment by rschlussel** [comment]
Looks like you still have changes that should be in the optimizers commit in the second commit instead.  Also, you should add martin aa a coauthor in the second commit.

**Comment by fgwang7w** [comment]
> Looks like you still have changes that should be in the optimizers commit in the second commit instead. Also, you should add martin aa a coauthor in the second commit.

I am not certain about your concern here, are you suggesting to separate out the changes in the analyzer/planner in the 2nd commit for (don't plan unnecessary sort)?

**Comment by aweisberg** [comment]
Was there code in this PR copied from another project? If so please follow the [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).

**Comment by rschlussel** [comment]
The commit "avoid planning unnecessary sort" also has the changes to add optimizers like "EvaluateZeroLimit",and the corresponding changes in the RemoveRedundantX optimizers.  Those changes belong in the first commit.

> > Looks like you still have changes that should be in the optimizers commit in the second commit instead. Also, you should add martin aa a coauthor in the second commit.
> 
> I am not certain about your concern here, are you suggesting to separate out the changes in the analyzer/planner in the 2nd commit for (don't plan unnecessary sort)?

**Comment by aweisberg** [comment]
The co-author tag in the second commit is wrong. Thanks!

**Comment by fgwang7w** [comment]
@aweisberg tag is fixed. 
@rschlussel reworked the code, there's been some works outside of cherrypicks, so I committed 2nd commit as a fix to 1st cherrypick commit. please help review if everything looks good. Thanks,

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroLimit.java [hunk 7]
```diff
@@ -23,6 +23,7 @@
 import static com.facebook.presto.sql.planner.plan.Patterns.Limit.count;
 import static com.facebook.presto.sql.planner.plan.Patterns.limit;
 
+@Deprecated
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroLimit.java]: shouldn't be deprecated

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java [hunk 8]
```diff
@@ -26,6 +26,7 @@
 /**
  * Replaces 0% sample node with empty values node.
  */
+@Deprecated
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java]: shouldn't be deprecated

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroDistinctLimit.java [hunk 9]
```diff
@@ -1,4 +1,3 @@
-
 /*
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroDistinctLimit.java:1]: deleting the newline should be part of the commit that adds this class

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java [hunk 10]
```diff
@@ -25,9 +25,8 @@
 
 /**
  * Replace TopN node
- * 1. With a Sort node when the subplan is guaranteed to produce fewer rows than N
- * 2. With its source when the subplan produces only one row
- * 3. With a empty ValuesNode when N is 0
+ * 1. With its source when the subplan is at most one row
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java:28]: this change should be part of the commit that adds this class

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantTopN.java [hunk 11]
```diff
@@ -81,6 +82,7 @@ public void testZeroTopN()
                                                 ImmutableList.of(
                                                         constantExpressions(BIGINT, 1L, 10L),
                                                         constantExpressions(BIGINT, 2L, 11L))))))
+                // TODO: verify contents
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantTopN.java:85]: what's with this todo

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantTopN.java [hunk 12]
```diff
@@ -70,6 +70,7 @@ public void test()
     public void testZeroTopN()
     {
         tester().assertThat(new EvaluateZeroTopN())
+                //tester().assertThat(new RemoveRedundantTopN())
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantTopN.java:73]: why is this commented out?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroLimit.java [hunk 13]
```diff
@@ -33,8 +33,7 @@ public void testDoesNotFire()
                 .on(p ->
                         p.limit(
                                 1,
-                                p.values(p.variable("a"))))
-                .doesNotFire();
+                                p.values(p.variable("a"))));
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroLimit.java:36]: can you share more about this change? Also, it should be in the previous commit.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroSample.java [hunk 14]
```diff
@@ -35,8 +35,7 @@ public void testDoesNotFire()
                         p.sample(
                                 0.15,
                                 Type.BERNOULLI,
-                                p.values(p.variable("a"))))
-                .doesNotFire();
+                                p.values(p.variable("a"))));
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroSample.java:38]: This belongs in the previous commit.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantLimit.java [hunk 15]
```diff
@@ -79,7 +79,7 @@ public void testForZeroLimit()
     @Test
     public void doesNotFire()
     {
-        tester().assertThat(new RemoveRedundantLimit())
+        tester().assertThat(new EvaluateZeroLimit())
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantLimit.java:82]: this should be in the previous commit and in the evaluateZeroLimit test

**Review by rschlussel — COMMENTED** [review]
Please squash the first two commits together.  Also, I know I was the one who requested separating out the zero limit/distinctlimit/topn into separate optimizers, but I didn't realize at the time that those were zero *Because of* the limit.  Now that i understand that, I think it would be fine to keep them in a single optimizer as it was originally.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroDistinctLimit.java (hunk 9 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroDistinctLimit.java:1]: my apology could you please elaborate a little bit? I am unsure what you mean by this "deleting this newline?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantTopN.java (hunk 12 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantTopN.java:73]: comment out the incorrect class to test, remove comment

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroLimit.java (hunk 13 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroLimit.java:36]: this was left out unintensionally. added back.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroDistinctLimit.java (hunk 9 above)
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroDistinctLimit.java:1]: If you look at the changes commit by commit, you'll see that the commit "Avoid planning unnecessary sort" has a change that removes an extra line at the start of EvaluteZeroLimit.java.  This should be done in the previous commit where the class is added (i.e. don't have the extra line in the first place)

Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroDistinctLimit.java:1]: understood, fixed now in first commit

**Comment by fgwang7w** [comment]
@rschlussel comments are revised and addressed. could you please give another review? Many thanks!

**Comment by fgwang7w** [comment]
@rschlussel could you please help review this PR again? all fix are in right commit version now. many thanks!

**Comment by fgwang7w** [comment]
@rschlussel could you please help approve this fix to be merged? This is a good performance fix to be included if possible soon. many thanks for help!

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java [hunk 16]
```diff
@@ -26,6 +26,7 @@
 /**
  * Replaces 0% sample node with empty values node.
  */
+
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java:29]: nit: remove this change.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 17]
```diff
@@ -170,6 +170,7 @@
     public static final String INLINE_SQL_FUNCTIONS = "inline_sql_functions";
     public static final String REMOTE_FUNCTIONS_ENABLED = "remote_functions_enabled";
     public static final String CHECK_ACCESS_CONTROL_ON_UTILIZED_COLUMNS_ONLY = "check_access_control_on_utilized_columns_only";
+    public static final String SKIP_REDUNDANT_SORT = "remove_redundant_sort";
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:173]: the string should be skip_redundant_sort to match the config property and variable name.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 18]
```diff
@@ -186,6 +186,20 @@ public PlanOptimizers(
                 taskCountEstimator);
     }
 
+    @PostConstruct
+    public void initialize()
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: this should go in the previous commit (undoing the accidental formatting changes you made there).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java [hunk 19]
```diff
@@ -261,6 +253,14 @@ private Patterns() {}
         }
     }
 
+    public static class DistinctLimit
```
Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java:256]: this belongs in the previous commit where DistinctLimit was introduced.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java [hunk 20]
```diff
@@ -241,12 +240,6 @@ public void testUndistributedOrderBy()
         assertQueryOrdered(undistributedOrderBy, "SELECT orderstatus FROM orders ORDER BY orderstatus");
     }
 
-    @Test
-    public void testOrderLimitCompaction()
-    {
-        assertQueryOrdered("SELECT * FROM (SELECT * FROM orders ORDER BY orderkey) LIMIT 10");
```
Comment (rschlussel) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java:247]: why is this test removed?

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java [hunk 21]
```diff
@@ -80,7 +80,6 @@ public void testOrderByWithOutputColumnReference()
         assertQueryOrdered("SELECT max(a) FROM (values (1,2), (2,1)) t(a,b) GROUP BY b ORDER BY b", "VALUES 2, 1");
         assertQueryOrdered("SELECT max(a) FROM (values (1,2), (2,1)) t(a,b) GROUP BY t.b ORDER BY t.b*1.0", "VALUES 2, 1");
         assertQueryOrdered("SELECT -(a+b) AS a, -(a+b) AS b, a+b FROM (values (41, 42), (-41, -42)) t(a,b) GROUP BY a+b ORDER BY a+b", "VALUES (-83, -83, 83), (83, 83, -83)");
-        assertQueryOrdered("SELECT c.a FROM (SELECT CAST(ROW(-a.a) AS ROW(a BIGINT)) a FROM (VALUES (2), (1)) a(a) GROUP BY a.a ORDER BY a.a) t(c)", "VALUES -2, -1");
```
Comment (rschlussel) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java:83]: why is this test removed?

**Review by rschlussel — COMMENTED** [review]
The title for the first commit is too long.  Shorten it to "Avoid planning unnecessary LIMIT/TopN/Sort/DistinctLimit", and then add more description in the commit message body.  We generally follow these guidelines: https://chris.beams.io/posts/git-commit/

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java (hunk 16 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java:29]: ok

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java (hunk 10 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantTopN.java:28]: ok will move to the 2nd commit, along with its corresponding test

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java (hunk 16 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EvaluateZeroSample.java:29]: removed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantLimit.java (hunk 15 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestRemoveRedundantLimit.java:82]: yes fixed

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java (hunk 17 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java:173]: yes they should be the same,fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 18 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: there's an issue to move these two methods, fixed

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java (hunk 19 above)
Comment (fgwang7w) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/Patterns.java:256]: fixed

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java (hunk 21 above)
Comment (fgwang7w) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java:83]: should not have been removed, I am rechecking in the right 2nd commit, fixed

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java (hunk 20 above)
Comment (fgwang7w) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestOrderByQueries.java:247]: should not be removed, fixed

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroSample.java (hunk 14 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestEvaluateZeroSample.java:38]: fixed

**Comment by fgwang7w** [comment]
Hi @rschlussel I revised both commits, regarding the last test removed, I have revised the testcase. Basically the 2nd commit introduces less preserved ordering in the subquery in which only combination of `ORDER BY`and `LIMIT` changes the specific order and set of rows, otherwise it would remain as unordered rows per SQL standard. @kaikalur please FYI

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 22]
```diff
@@ -3389,9 +3389,12 @@ public void testCorrelatedScalarSubqueries()
         // two level of nesting
         assertQuery("SELECT * FROM nation n WHERE 2 = (SELECT (SELECT 2 * n.nationkey))");
 
+        // redundant LIMIT in subquery
+        assertQuery("SELECT (SELECT count(*) FROM (VALUES (7,1)) t(orderkey, value) WHERE orderkey = corr_key LIMIT 1) FROM (values 7) t(corr_key)");
+
         // explicit LIMIT in subquery
         assertQueryFails(
-                "SELECT (SELECT count(*) FROM (VALUES (7,1)) t(orderkey, value) WHERE orderkey = corr_key LIMIT 1) FROM (values 7) t(corr_key)",
+                "SELECT (SELECT count(*) FROM (VALUES (7,1)) t(orderkey, value) WHERE orderkey = corr_key GROUP BY value LIMIT 1) FROM (values 7) t(corr_key)",
```
Comment (rschlussel) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:3397]: It seems in this case the limit is also redundant.  We're just not smart enough to remove it.  Maybe instead have 2 rows in the values node.

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java [hunk 23]
```diff
@@ -86,16 +86,16 @@ public void testCorrelatedExistsSubqueriesWithOrPredicateAndNull()
     }
 
     @Test
-    public void testUnsupportedSubqueriesWithCoercions()
```
Comment (rschlussel) [inline presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:89]: I think limit here was incidental and we still want this test for unsupported subqueries to ensure they throw proper errors.

**Review by rschlussel — COMMENTED** [review]
Nearly there. just some small comments about the tests.

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 22 above)
Comment (fgwang7w) [reply presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java:3397]: or I can change it to an upper bound of LIMIT, then it will make sense,but this is a good catch, we should remove limit if there's a enforceSingleRow like count(*) to make the optimizer smarter. I propose we fix it in a separate PR so that `TransformCorrelatedSingleRowSubqueryToProject` can take it into account

File: presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java (hunk 23 above)
Comment (fgwang7w) [reply presto-main/src/test/java/com/facebook/presto/sql/query/TestSubqueries.java:89]: then I will add a seperate tc with `group by a limit 1` to guard this sanity test

**Review by rschlussel — APPROVED** [review]
Looks good. Thanks for sticking with it!

**Comment by kaikalur** [comment]
I'm not sure if we should hav those plan tests. We are planning to do more rewrites/optimizer rules so it's going to be painful to test these in the future. I say the plan tests should not be included.

**Comment by rschlussel** [comment]
> I'm not sure if we should hav those plan tests. We are planning to do more rewrites/optimizer rules so it's going to be painful to test these in the future. I say the plan tests should not be included. @rongrong WDYT?

Which plan tests do you mean?  If you mean the tpch plan tests, those already exist, so if you want to get rid of them, i think that belongs in a separate pr. (I think the purpose of them is just to let you know if you are altering tpch or tpcds plans, so that you can benchmark whether there was a regression).

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 24]
```diff
@@ -1122,12 +1127,22 @@ protected Scope visitQuerySpecification(QuerySpecification node, Optional<Scope>
...
-            else {
-                analysis.setOrderByExpressions(node, emptyList());
+                if (node.getSelect().isDistinct()) {
+                    verifySelectDistinct(node, outputExpressions);
+                }
+
+                OrderBy orderBy = node.getOrderBy().get();
+                orderByScope = Optional.of(computeAndAssignOrderByScope(orderBy, sourceScope, outputScope));
+
+                orderByExpressions = analyzeOrderBy(node, orderBy.getSortItems(), orderByScope.get());
```
[the remark this answers is not in the document]
Comment (rschlussel) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1137]: this is the analyzer?

[the remark this answers is not in the document]
Comment (kaikalur) [reply presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java:1137]: > this is the analyzer?

Oops - I think I got lost in the see of files in this large PR :)

**Review by kaikalur — APPROVED** [review]
LGTM
